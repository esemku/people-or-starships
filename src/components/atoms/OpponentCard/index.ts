export { default as OpponentCard } from './OpponentCard';
