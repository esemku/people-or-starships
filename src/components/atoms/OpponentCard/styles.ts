import { Theme } from '@mui/material';
import { makeStyles } from '@mui/styles';

export default makeStyles((theme: Theme) => ({
  root: {
    width: 400,
    marginLeft: 20,
    marginRight: 20,
    position: 'relative',
    overflow: 'visible !important',
  },
  rootWinner: {
    border: `1px solid ${theme.palette.success.light}`,
  },
  winnerText: {
    position: 'absolute',
    top: '-35px',
    left: '50%',
    transform: 'translate(-50%, 0)',
    fontSize: 21,
    color: theme.palette.success.light,
  },
}));
