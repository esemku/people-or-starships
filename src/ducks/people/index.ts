export {
  PEOPLE_ACTION_TYPES,
  loadPeople,
  loadPeopleSuccess,
  loadPeopleFailure,
} from './actions';
export { default } from './reducers';
export { getPeople } from './selectors';
