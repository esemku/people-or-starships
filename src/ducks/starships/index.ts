export {
  STARSHIPS_ACTION_TYPES,
  loadStarships,
  loadStarshipsSuccess,
  loadStarshipsFailure,
} from './actions';
export { default } from './reducers';
export { getStarships } from './selectors';
