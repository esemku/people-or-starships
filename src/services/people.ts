import client from 'utils/client';

export const getPeople = async () => {
  const response = await client({
    url: `/people`,
    method: 'GET',
  });
  return response.data;
};
