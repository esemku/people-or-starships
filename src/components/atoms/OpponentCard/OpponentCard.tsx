import React from 'react';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import Typography from '@mui/material/Typography';
import { Person } from 'types/people';
import { OpponentsKind } from 'components/App';
import { PEOPLE_CARD_ITEMS, STARSHIPS_CARD_ITEMS } from 'utils/cardItems';
import { Starship } from 'types/starships';
import clsx from 'clsx';
import useStyles from './styles';

type OpponentCardProps = {
  opponent: Person | Starship;
  isWinner: boolean;
  opponentsKind: OpponentsKind;
};

const OpponentCard: React.FC<OpponentCardProps> = ({
  opponent,
  isWinner,
  opponentsKind,
}: OpponentCardProps) => {
  const styles = useStyles();

  return (
    <Card className={clsx(styles.root, isWinner && styles.rootWinner)}>
      {isWinner && <div className={styles.winnerText}>WINNER!</div>}
      <CardContent>
        {(opponentsKind === 'people'
          ? PEOPLE_CARD_ITEMS
          : STARSHIPS_CARD_ITEMS
        ).map((cardItem) => {
          if (cardItem.parameter === 'name') {
            return (
              <Typography variant="body1" gutterBottom noWrap>
                {`${cardItem.label}: ${opponent[cardItem.parameter]}`}
              </Typography>
            );
          }
          return (
            <Typography
              variant="body2"
              color="text.secondary"
              gutterBottom
              noWrap
            >
              {`${cardItem.label}: ${opponent[cardItem.parameter]}`}
            </Typography>
          );
        })}
      </CardContent>
    </Card>
  );
};

export default OpponentCard;
