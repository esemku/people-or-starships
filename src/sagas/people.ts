import {
  loadPeopleFailure,
  loadPeopleSuccess,
  PEOPLE_ACTION_TYPES,
} from 'ducks/people';
import { ActionLoadPeople } from 'ducks/people/actions';
import {
  all,
  call,
  takeLatest,
  put,
  CallEffect,
  PutEffect,
} from 'redux-saga/effects';
import { PeopleService } from 'services';
import { Person } from 'types/people';

export type LoadPeopleResponse = {
  count: number;
  next: string;
  previous: string;
  results: Person[];
};

export function* loadPeopleSaga(): Generator<
  CallEffect<LoadPeopleResponse> | PutEffect<ActionLoadPeople>,
  void,
  LoadPeopleResponse
> {
  const response: LoadPeopleResponse = yield call(PeopleService.getPeople);

  if (response) {
    yield put(loadPeopleSuccess(response.results));
  } else {
    yield put(loadPeopleFailure());
  }
}

function* peopleSaga() {
  yield all([
    yield takeLatest(PEOPLE_ACTION_TYPES.LOAD_PEOPLE, loadPeopleSaga),
  ]);
}

export default peopleSaga;
