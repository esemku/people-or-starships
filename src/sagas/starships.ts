import {
  loadStarshipsFailure,
  loadStarshipsSuccess,
  STARSHIPS_ACTION_TYPES,
} from 'ducks/starships';
import { ActionLoadStarships } from 'ducks/starships/actions';
import {
  all,
  call,
  takeLatest,
  put,
  CallEffect,
  PutEffect,
} from 'redux-saga/effects';
import { StarshipsService } from 'services';
import { Starship } from 'types/starships';

export type LoadStarshipsResponse = {
  count: number;
  next: string;
  previous: string;
  results: Starship[];
};

export function* loadStarshipsSaga(): Generator<
  CallEffect<LoadStarshipsResponse> | PutEffect<ActionLoadStarships>,
  void,
  LoadStarshipsResponse
> {
  try {
    const response: LoadStarshipsResponse = yield call(
      StarshipsService.getStarships,
    );

    if (response) {
      yield put(loadStarshipsSuccess(response.results));
    } else {
      yield put(loadStarshipsFailure());
    }
  } catch (e) {
    yield put(loadStarshipsFailure());
  }
}

function* starshipsSaga() {
  yield all([
    yield takeLatest(STARSHIPS_ACTION_TYPES.LOAD_STARSHIPS, loadStarshipsSaga),
  ]);
}

export default starshipsSaga;
