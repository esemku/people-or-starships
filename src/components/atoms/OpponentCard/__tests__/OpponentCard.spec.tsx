import React from 'react';
import { render, screen } from '@testing-library/react';

import { PersonMock } from 'mocks/people.mock';
import { ThemeProvider } from '@mui/material';
import theme from 'theme';
import { StarshipMock } from 'mocks/starships.mock';
import OpponentCard from '../OpponentCard';

describe('<OpponentCard />', () => {
  it('renders correctly', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <OpponentCard
          opponent={PersonMock}
          isWinner={false}
          opponentsKind="people"
        />
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
  });
  it('renders winnerText when isWinner is true', () => {
    render(
      <ThemeProvider theme={theme}>
        <OpponentCard
          opponent={StarshipMock}
          isWinner
          opponentsKind="starships"
        />
      </ThemeProvider>,
    );

    expect(screen.getByTestId('winnerText')).toBeInTheDocument();
  });
});
