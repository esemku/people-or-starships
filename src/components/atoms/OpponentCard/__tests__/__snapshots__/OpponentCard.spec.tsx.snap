// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OpponentCard /> renders correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root makeStyles-root-1 css-2bjnfd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiTypography-noWrap css-577owk-MuiTypography-root"
      >
        Name: R5-D4
      </p>
      <p
        class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom MuiTypography-noWrap css-tpcsis-MuiTypography-root"
      >
        Birth Year: unknown
      </p>
      <p
        class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom MuiTypography-noWrap css-tpcsis-MuiTypography-root"
      >
        Gender: n/a
      </p>
      <p
        class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom MuiTypography-noWrap css-tpcsis-MuiTypography-root"
      >
        Height: 97
      </p>
      <p
        class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom MuiTypography-noWrap css-tpcsis-MuiTypography-root"
      >
        Eye color: red
      </p>
    </div>
  </div>
</div>
`;
